### PR TITLE
Specify min/max versions for all dependencies in setup.py

### PR DESCRIPTION
**Pull request by zaneselvans**
At @ezwelty's suggestion (after some unexpected breakage) I've tried to set minimum and maximum versions for all of our direct dependencies in `setup.py` so that unexpected major breaking updates should be much less frequent. I'm not entirely sure how this will interact with our use of `environment.yml` to specify the `pudl-dev` environment, but so far it seems to work fine with `tox` and `pip`. Recreating my `pudl-dev` environment locally from scratch, and then installing pudl inside it using pip seems to work fine though.

### CODE REVIEW

**Comment by ezwelty** [comment]
@zaneselvans It will be tidier to use compatible release clauses (https://www.python.org/dev/peps/pep-0440/#compatible-release). For example:

- `>=1.2.0, <1.3.0` => `~=1.2.0`
- `>=1.2, <2.0` => `~=1.2`

Also, did you forget `flake8-rst-docstrings`? Otherwise, the version logic seems good to me.

File: setup.py [hunk 1]
```diff
@@ -7,61 +7,61 @@
...
-    "flake8-colors",
-    "flake8-docstrings",
+    "bandit>=1.6,<2.0",
+    "coverage>=5.3,<6.0",
+    "doc8>=0.8,<0.9",
+    "flake8>=3.8,<4.0",
+    "flake8-builtins>=1.5,<2.0",
+    "flake8-colors>=0.1,<0.2",
+    "flake8-docstrings>=1.5,<2.0",
     "flake8-rst-docstrings",
```
Comment (ezwelty) [inline setup.py:56]: Missing version constraints.

Comment (zaneselvans) [reply setup.py:56]: The most recent version was 0.0.14 and so I wasn't sure what to do with it! Restricting to <0.1.0 seemed a bit odd.

**Comment by ezwelty** [comment]
This looks like a good place to start, by at least protecting us from automatic upgrades most likely to introduce breaking changes.

I've been using [poetry](https://python-poetry.org) to manage dependencies in new python projects, and it uses a "lock file" with resolved dependency versions (which gets committed to the repository) so that all environments match exactly – essentially equivalent to pinning all versions explicitly, but with the benefit of having a layer of software figuring out which combination actually works. But since there are difficulties with using `poetry` + `conda`, I don't necessarily recommend it here.

It makes me nervous to have dependency versions specified twice – in `setup.py` and `environment.yml`. But a quick internet search did not yield easy answers on how to populate, say, `setup.py` from a YML file.

**Review by ezwelty — APPROVED** [review]

**Pull request merged**